src/lib/bakeryjs/tracingModel.ts

```diff
@@ -435,4 +414,41 @@ export class TracingModel {
 		);
 		return;
 	}
+
+	private getSubDimensionsDone(msgId: string) {
+		if (!this.dimensionStore.has(msgId)) return true;
+
+		const dimValues = this.dimensionStore.get(msgId).values();
+		return every(
+			[...dimValues],
+			(dt: DimensionTrace) => dt.complete && dt.done
+		);
+	}
+
+	private getDimensionDone(parentMsgId: string, dimension: string[]) {
+		if (!this.dimensionStore.get(parentMsgId).get(dimension).complete) {
+			return false;
+		}
+
+		const dimValues = this.msgStore
+			.get(parentMsgId)
+			.get(dimension)
+			.values();
+
+		return every([...dimValues], (mT: MsgTrace) => mT.done);
+	}
+
+	private getBoxesDone(
+		msgId: string,
+		parentMsgId: string,
+		dimension: string[]
+	) {
+		const dimValues = this.msgStore
+			.get(parentMsgId)
+			.get(dimension)
+			.get(msgId)
+			.boxes.values();
+
+		return every([...dimValues], Boolean);
```

Comment:
same here, no spreading.
Remember that spreading like this is expensive if very frequent or with huge arrays.